source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```diff
@@ -0,0 +1,419 @@
+            instance.Name = dimension.Name;
+            foreach (var value in dimension.Values)
+            {
+                instance.Values.Add(EncodePrimitiveValue(value));
+            }
+
+            return instance;
+        }
+
+        public OptimizerService.PrimitiveValue EncodePrimitiveValue(object value)
+        {
+            var data = new OptimizerService.PrimitiveValue();
+            if (value is int)
+            {
+                data.IntValue = (int)value;
+            }
+            else if (value is bool)
+            {
+                data.BoolValue = (bool)value;
+            }
+            else if (value is string)
+            {
+                data.StringValue = (string)value;
+            }
+            else if (value is double)
+            {
+                data.DoubleValue = (double)value;
+            }
+            else
+            {
+                System.Console.Error.WriteLine("Invalid type provided to OptimizerServiceEncoderDecoder.Encoder.EncodePrimitiveValue(..). Type: " + value.GetType());
+                data.IntValue = 0;
+            }
+
+            return data;
+        }
+    }
+
+    public class OptimizerServiceDecoder
+    {
+        private DimensionTypeName GrpcTypeToDimension(OptimizerService.DimensionType typeName)
+        {
+            switch (typeName)
+            {
+                case OptimizerService.DimensionType.Categorical:
+                    return DimensionTypeName.CategoricalDimension;
+                case OptimizerService.DimensionType.Continuous:
+                    return DimensionTypeName.ContinuousDimension;
+                case OptimizerService.DimensionType.Ordinal:
+                    return DimensionTypeName.OrdinalDimension;
+                case OptimizerService.DimensionType.Discrete:
+                    return DimensionTypeName.DiscreteDimension;
+                default:
+                    System.Console.Error.WriteLine("Invalid type provided to OptimizerServiceEncoderDecoder.Decoder.GrpcTypeToDimension(..). Type: " + typeName);
+
+                    return DimensionTypeName.ContinuousDimension;
+            }
+        }
+
+        public OptimizationProblem DecodeOptimizationProblem(OptimizerService.OptimizationProblem problem)
+        {
+            var objectives = new OptimizationObjective[problem.Objectives.Count];
+            var i = 0;
+            foreach (var objective in problem.Objectives)
+            {
+                objectives[i++] = DecodeOptimizationObjective(objective);
+            }
+
+            var instance = new OptimizationProblem(
+                        DecodeHypergrid(problem.ParameterSpace),
+                        DecodeHypergrid(problem.ObjectiveSpace),
+                        objectives);
+            if (problem.ContextOptionsCase == OptimizerService.OptimizationProblem.ContextOptionsOneofCase.ContextSpace)
+            {
+                // A context space was provided

```

Comment:
I think in C# we usually write comments with an extra empty comment line:
```suggestion
                // A context space was provided
                //

```